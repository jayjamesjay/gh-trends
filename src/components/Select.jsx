import React from 'react';
import PropTypes from 'prop-types';
import StyledSelect from '../styles/Select';
import Option from '../styles/Option';
import { ImgIcon } from '../styles/Img';
import FilterImg from '../assets/img/filter-list.svg';

/**
 * @module Select
 */

/**
 * Language selection
 *
 * @property {string} curr - currently selected language
 * @property {function} onSelect - action fired when language is selected
 * @property {array} languages - list of languages to select from
 * @property {string} label - description of this language selection
 * @returns {SelectLang}
 */
export function SelectLang(props) {
  const { options, label } = props;
  const langs = ['All Languages'].concat(options);

  return (
    <>
      <ImgIcon src={FilterImg} alt={label} title={label} />
      <Select {...props} options={langs} label="Languages" />
    </>
  );
}

SelectLang.propTypes = {
  curr: PropTypes.string.isRequired,
  onSelect: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(PropTypes.string).isRequired,
  label: PropTypes.string.isRequired
};

/**
 * General selection
 *
 * @property {string} curr - currently selected option
 * @property {function} onSelect - action fired when option is selected
 * @property {array} options - list of options to select from
 * @property {string} label - description of this language selection
 * @returns {Select}
 */
export default function Select({ curr, onSelect, options, label }) {
  return (
    <StyledSelect value={curr} onChange={onSelect} aria-label={label}>
      {options.map(name => (
        <Option key={name} value={name}>
          {name}
        </Option>
      ))}
    </StyledSelect>
  );
}

Select.propTypes = {
  curr: PropTypes.string.isRequired,
  onSelect: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(PropTypes.string).isRequired,
  label: PropTypes.string.isRequired
};
