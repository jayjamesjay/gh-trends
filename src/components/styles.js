import { Link as RouterLink } from "react-router-dom";
import styled from "styled-components";
import { createGlobalStyle } from "styled-components";
import { light } from "./theme";

export const GlobalStyle = createGlobalStyle`
   body {
     margin: 0;
     font-family: "Noto Sans HK", sans-serif;
     background: ${props => props.theme.bg};
   }
`;

export const Header = styled.header`
  display: flex;
  justify-content: space-between;
  text-transform: uppercase;
  position: sticky;
  top: 0;
  background: ${props => props.theme.bgAdd};
  box-shadow: 0 1px 2px 0 rgba(60, 64, 67, 0.3),
    0 2px 6px 2px rgba(60, 64, 67, 0.15);
  padding: env(safe-area-inset-top) env(safe-area-inset-right)
    env(safe-area-inset-bottom) env(safe-area-inset-left);
`;

Header.defaultProps = {
  theme: light
};

export const HeaderTitle = styled.h1`
  font-size: 2rem;
  margin: 0.5rem;
`;

export const Link = styled(RouterLink)`
  color: ${props => props.theme.color};
  text-decoration: none;
  font-size: ${props => props.fontSize};
  font-family: "Noto Sans HK", sans-serif;
  padding: 0.5rem;
`;

Link.defaultProps = {
  theme: light,
  fontSize: "1.3rem"
};

export const Nav = styled.nav`
  transition: opacity 0.2s ease-in-out;
  opacity: ${props => (props.hide ? "0" : "1.0")};
  position: fixed;
  top: 3.9rem;
  left: 0;
  z-index: 100;
  width: 100%;
  background: ${props => props.theme.bgAdd};
  box-shadow: 0 3px 6px 0 rgba(104, 104, 104, 0.5);
  padding: 0.7rem 0;

  @media screen and (min-width: 40rem) {
    top: 0.4rem;
    left: 47%;
    width: 50%;
    opacity: 1;
    box-shadow: none;
  }
`;

Nav.defaultProps = {
  theme: light,
  hide: true
};

export const Menu = styled.ul`
  display: flex;
  flex-flow: row wrap;
  justify-content: space-around;
  text-transform: uppercase;
  list-style-type: none;
  margin: 0;
  padding: 0;

  @media screen and (min-width: 40rem) {
    justify-content: flex-end;
  }
`;

export const MenuToggle = styled.button`
  overflow: hidden;
  cursor: pointer;
  background: transparent;
  margin: 0 1rem;
  padding: 0;
  border: 0;

  @media screen and (min-width: 40rem) {
    display: none;
  }

  & > span {
    transition: transform 0.2s linear;
    display: block;
    background: ${props => props.theme.color};
    width: 1.8rem;
    height: 0.2rem;
    margin: 0.3rem 0;

    &:first-child {
      transform: ${props =>
        props.open ? "rotate(-45deg) translate(-0.35rem, 0.35rem)" : ""};
    }

    &:nth-child(2) {
      transform: ${props => (props.open ? "translateX(15rem)" : "")};
    }

    &:last-child {
      transform: ${props =>
        props.open ? "rotate(45deg) translate(-0.35rem, -0.35rem)" : ""};
    }
  }
`;

MenuToggle.defaultProps = {
  theme: light,
  open: true
};

export const Main = styled.main`
  padding: env(safe-area-inset-top) env(safe-area-inset-right)
    env(safe-area-inset-bottom) env(safe-area-inset-left);
  background: ${props => props.theme.bg};
  color: ${props => props.theme.color};

  @media screen and (min-width: 40rem) {
    display: flex;
    flex-flow: row wrap;
    justify-content: space-around;
  }
`;

Main.defaultProps = {
  theme: light
};

export const Article = styled.article`
  background: ${props => props.theme.bgAdd};
  background: linear-gradient(
    139deg,
    ${props => props.theme.bgAdd} 0%,
    ${props => props.theme.bgAdd} 57%,
    ${props => props.bg} 57%
  );
  color: ${props => props.theme.color};
  padding: 0.7rem 1.2rem;
  margin: 1rem;
  box-shadow: 0 1px 2px 0 rgba(60, 64, 67, 0.3);
  min-height: 21rem;

  @media screen and (min-width: 40rem) {
    margin: 2rem;
    min-height: 10rem;
    flex: 1 1 30%;
  }
`;

Article.defaultProps = {
  theme: light,
  bg: "#aaa"
};

export const H2 = styled.h2`
  font-size: 1.7rem;
`;

export const Span = styled.span`
  color: ${props => props.theme.colorAdd};
  font-size: 1.3rem;
  font-weight: 400;
`;

Span.defaultProps = {
  theme: light
};

export const List = styled.ul`
  display: flex;
  flex-flow: row wrap;
  margin: 0;
  padding: 0;
  width: 60%;
  list-style-type: none;
`;

export const ListItem = styled.li`
  color: ${props => props.theme.colorAdd};
  padding: 0.5rem 1rem 0 0;
`;

export const P = styled.p`
  width: 65%;
`;

export const Pr = styled.p`
  text-align: right;
`;

export const Footer = styled.footer`
  padding: 0.5rem;
  text-align: center;
  background: ${props => props.theme.bg};
  color: ${props => props.theme.color};
`;

Footer.defaultProps = {
  theme: light
};
