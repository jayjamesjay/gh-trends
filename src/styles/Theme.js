export const light = {
  bg: '#fafafa',
  bgAdd: '#fff',
  shadow: '0px 3px 5px 1px rgba(217,217,217,0.82)',
  color: '#000',
  colorAdd: '#767676',
};

export const dark = {
  bg: '#000',
  bgAdd: '#202124',
  shadow: 'none',
  color: '#fff',
  colorAdd: '#8d8d8d',
};
