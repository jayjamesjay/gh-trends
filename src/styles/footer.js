import styled from 'styled-components';

export const Footer = styled.footer`
  margin: 1rem;
  padding: 0.5rem;
  text-align: center;
`;

export default Footer;
