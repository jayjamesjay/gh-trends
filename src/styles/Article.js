import styled from 'styled-components';
import { light } from './Theme';

export const oppositeBg = props => (props.bg ? props.bg : props.theme.color);

const Article = styled.article`
  display: flex;
  margin: 2rem 1rem;
  min-height: 13rem;
  padding: 1.2rem 1.2rem 0.8rem 1.2rem;
  flex-flow: column nowrap;
  background: linear-gradient(
    ${props => props.theme.bgAdd} 0%,
    ${props => props.theme.bgAdd} 0.3rem,
    ${props => oppositeBg(props)} 0.3rem,
    ${props => oppositeBg(props)} 0.6rem,
    ${props => props.theme.bgAdd} 0.6rem
  );
  box-shadow: ${props => props.theme.shadow};
  border-radius: 0.5rem;
  text-align: left;
  word-break: break-word;

  @media screen and (min-width: 40rem) {
    min-width: 16.5rem;
    margin: 1.5rem 1rem;
    flex: 1 0 27%;
  }
`;

Article.defaultProps = {
  theme: light
};

export default Article;
