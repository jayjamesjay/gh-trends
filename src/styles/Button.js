import styled from 'styled-components';
import { light } from './Theme';

export const Button = styled.button`
  cursor: pointer;
  background: ${(props) => props.theme.bgAdd};
  border: 0;
  border-radius: 0.5rem;
  font-size: 1.2rem;
  color: ${(props) => props.theme.color};
  text-transform: uppercase;
`;

Button.defaultProps = {
  theme: light,
};

export const ButtonTheme = styled(Button)`
  width: 3rem;
  height: 3rem;
  margin: 0 auto;
  background: transparent;

  @media screen and (min-width: 40rem) {
    margin: 0 1rem;
  }
`;

export const ButtonMain = styled(Button)`
  margin: auto;
  padding: ${(props) => (props.$visible ? '1rem' : '0')};
  box-shadow: ${(props) => props.theme.shadow};
  width: ${(props) => (props.$visible ? 'auto' : '0')};
  height: ${(props) => (props.$visible ? 'auto' : '0')};
  overflow: ${(props) => (props.$visible ? '' : 'hidden')};
`;

ButtonMain.defaultProps = {
  theme: light,
};

export const ButtonIcon = styled.button`
  cursor: pointer;
  width: 2rem;
  height: 2rem;
  background: transparent;
  border: 0;
`;

export const ButtonRemove = styled(ButtonIcon)`
  margin-top: -0.6rem;
`;

export const ButtonAdd = styled(ButtonIcon)`
  margin: 0.5rem 0 0 0;
`;

export const MenuToggle = styled(Button)`
  margin: 0 1rem;
  padding: 0;
  background: transparent;
  overflow: hidden;

  @media screen and (min-width: 40rem) {
    display: none;
  }
`;

MenuToggle.defaultProps = {
  theme: light,
};

export const ToggleSpan = styled.span`
  display: block;
  width: 1.8rem;
  height: 0.2rem;
  margin: 0.3rem 0;
  background: ${(props) => props.theme.color};
  transition: transform 0.2s linear;

  &:first-child {
    transform: ${(props) => (props.$open ? 'translateY(0.45rem) rotate(-45deg)' : '')};
  }

  &:nth-child(2) {
    transform: ${(props) => (props.$open ? 'translateX(5rem)' : '')};
  }

  &:last-child {
    transform: ${(props) => (props.$open ? 'translateY(-0.45rem) rotate(45deg)' : '')};
  }
`;

ToggleSpan.defaultProps = {
  theme: light,
  open: true,
};
