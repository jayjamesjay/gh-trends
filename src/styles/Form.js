import styled from 'styled-components';
import { light } from './Theme';

export const TextInput = styled.input`
  background: ${props => props.theme.bg};
  color: ${props => props.theme.color};
  padding: 0.5rem;
  font-size: 1.2rem;
  border: none;
  border: solid ${props => props.theme.color};
  border-width: 0 0 0.15rem 0;
  min-width: 50%;

  &:focus {
    border-width: 0.15rem;
  }
`;

export const Form = styled.form`
  display: flex;
  flex-flow: row wrap;
  justify-content: center;
`;

export const FormAlt = styled(Form)`
  padding: 1rem;
`;

export const Fieldset = styled.fieldset`
  padding: 1rem;
  flex: 1 1 100%;
  border: none;
  display: flex;
  flex-flow: row wrap;
  justify-content: center;
`;

export const CategoryMenu = styled.form`
  display: flex;
  flex-flow: row wrap;
  text-transform: uppercase;
  width: 100%;
  justify-content: center;
`;

export const InputRadio = styled.input`
  opacity: 0;
`;

export const LabelTab = styled.label`
  display: inherit;
  cursor: pointer;
  font-size: 1.3rem;
  padding: 1rem 1.5rem;
  background: transparent;
  position: relative;
  overflow: hidden;

  &::before {
    transition: transform 0.15s ease-in-out;
    position: absolute;
    z-index: 20;
    top: 5%;
    left: 5%;
    content: '';
    width: 90%;
    height: 90%;
    transform: translateX(${props => (props.active ? '0%' : '-200%')});
    background: ${props => props.theme.bgAdd};
    box-shadow: ${props => props.theme.shadow};
  }
`;

export const LabelSpan = styled.span`
  position: relative;
  z-index: 100;
`;

LabelTab.defaultProps = {
  theme: light
};

export const LabelSelect = styled(LabelTab)`
  display: flex;
  flex-flow: row wrap;
  align-content: center;
  align-items: center;
  justify-content: center;
  margin: 2rem auto;
  max-width: 80%;
  padding: 0.7rem 2.5rem;
  text-transform: uppercase;

  @media screen and (min-width: 30rem) {
    max-width: 28rem;
    flex: 1 1 100%;
  }
`;

export const Select = styled.select`
  background: ${props => props.theme.bgAdd};
  color: ${props => props.theme.color};
  box-shadow: ${props => props.theme.shadow};
  border: none;
  padding: 0.5rem 0;
  margin: 0.5rem;
  font-size: 1.3rem;
  appearance: none;
  text-transform: uppercase;
  text-align: center;

  @media screen and (min-width: 40rem) {
    margin: 0.15rem 0 0 1rem;
  }
`;

export const Option = styled.option`
  font-size: 1.2rem;
  text-transform: uppercase;
  line-break: loose;
  text-align: center;
`;
