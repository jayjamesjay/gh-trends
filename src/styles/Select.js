import styled from 'styled-components';
import { light } from './Theme';

const Select = styled.select`
  max-width: 80%;
  margin: 0.5rem;
  padding: 0.5rem;
  background: ${props => props.theme.bgAdd};
  border: none;
  box-shadow: ${props => props.theme.shadow};
  font-size: 1.3rem;
  color: ${props => props.theme.color};
  text-align: center;
  text-transform: uppercase;
  appearance: none;
`;

Select.defaultProps = {
  theme: light
};

export default Select;
