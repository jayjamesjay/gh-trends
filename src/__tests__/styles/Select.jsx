import React from 'react';
import 'jest-styled-components';
import { render } from '@testing-library/react';
import Select from '../../styles/Select';
import { dark } from '../../styles/Theme';

describe('<Select />', () => {
  it('renders with default style', () => {
    const { asFragment } = render(<Select />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with theme', () => {
    const { asFragment } = render(<Select theme={dark} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
