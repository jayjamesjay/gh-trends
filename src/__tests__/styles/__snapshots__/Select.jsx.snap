// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> renders with default style 1`] = `
<DocumentFragment>
  .c0 {
  max-width: 80%;
  margin: 0.5rem;
  padding: 0.5rem;
  background: #fff;
  border: none;
  border-radius: 0.5rem;
  box-shadow: 0px 3px 5px 1px rgba(217,217,217,0.82);
  font-family: 'Noto Sans HK',sans-serif;
  font-size: 1.3rem;
  color: #000;
  text-align: center;
  text-transform: uppercase;
  appearance: none;
}

<select
    class="c0"
  />
</DocumentFragment>
`;

exports[`<Select /> renders with theme 1`] = `
<DocumentFragment>
  .c0 {
  max-width: 80%;
  margin: 0.5rem;
  padding: 0.5rem;
  background: #202124;
  border: none;
  border-radius: 0.5rem;
  box-shadow: none;
  font-family: 'Noto Sans HK',sans-serif;
  font-size: 1.3rem;
  color: #fff;
  text-align: center;
  text-transform: uppercase;
  appearance: none;
}

<select
    class="c0"
  />
</DocumentFragment>
`;
